### PR TITLE
Feature/move lambda parsing

### PR DESCRIPTION
**Pull request by garyluu**
Moved the language parsing lambda back to this repo

Using Google's codestyle with Google (with Dockstore additions) checkstyle

The openapi.yaml URL needs to be replaced once the web service portion is merged

Also considering to remove the README since it's generated and unlikely to be maintained

### CODE REVIEW

File: .github/workflows/maven.yml [hunk 1]
```diff
@@ -32,5 +32,11 @@ jobs:
        run: | 
          sam local invoke HelloWorldFunction -e events/event.json -n env.json &> output.txt
          grep "Valid push event" output.txt
+     - name: Test wdl parsing with SAM CLI build and invoke
```
Comment (denis-yuen) [inline .github/workflows/maven.yml:35]: Nice

File: wdl-parsing/WDLParsingFunction/ignore.xml [hunk 2]
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<FindBugsFilter>
```
Comment (denis-yuen) [inline wdl-parsing/WDLParsingFunction/ignore.xml:2]: code comment, why is this needed and why is it ok?

File: wdl-parsing/WDLParsingFunction/pom.xml [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+    <modelVersion>4.0.0</modelVersion>
+    <groupId>io.dockstore</groupId>
+    <artifactId>wdl-parser</artifactId>
+    <version>1.0</version>
+    <packaging>jar</packaging>
+    <name>WDL parser created for SAM CLI.</name>
+    <properties>
+        <maven.compiler.source>11</maven.compiler.source>
+        <maven.compiler.target>11</maven.compiler.target>
+        <cromwell.version>51</cromwell.version>
```
Comment (denis-yuen) [inline wdl-parsing/WDLParsingFunction/pom.xml:12]: fairly old version of cromwell, annotations
maybe check the other stuff too

File: wdl-parsing/WDLParsingFunction/src/main/java/io/dockstore/wdlparser/package-info.java [hunk 4]
```diff
@@ -0,0 +1,16 @@
+/*
+ *    Copyright 2020 OICR
```
Comment (denis-yuen) [inline wdl-parsing/WDLParsingFunction/src/main/java/io/dockstore/wdlparser/package-info.java:2]: gotta iterate the date and add uc?

File: wdl-parsing/WDLParsingFunction/checkstyle.xml [hunk 5]
```diff
@@ -0,0 +1,424 @@
+<?xml version="1.0"?>
+<!DOCTYPE module PUBLIC
+        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
+        "https://checkstyle.org/dtds/configuration_1_3.dtd">
+
+<!--
```
[the remark this answers is not in the document]
Comment (garyluu) [reply wdl-parsing/WDLParsingFunction/checkstyle.xml:6]: I'll add a comment in both checkstyle files, but I think this should be the gold copy since it's more strict and fully compatible with the Google codestyle (instead of the dockstore/dockstore-modified one)

File: wdl-parsing/WDLParsingFunction/ignore.xml (hunk 2 above)
Comment (garyluu) [reply wdl-parsing/WDLParsingFunction/ignore.xml:2]: I'll add a code comment, but it appears it's because they're generated files

File: wdl-parsing/WDLParsingFunction/src/main/java/io/dockstore/wdlparser/package-info.java (hunk 4 above)
Comment (garyluu) [reply wdl-parsing/WDLParsingFunction/src/main/java/io/dockstore/wdlparser/package-info.java:2]: example?

File: wdl-parsing/WDLParsingFunction/checkstyle.xml (hunk 5 above)
[the remark this answers is not in the document]
Comment (denis-yuen) [reply wdl-parsing/WDLParsingFunction/checkstyle.xml:6]: That works too

File: wdl-parsing/WDLParsingFunction/src/main/java/io/dockstore/wdlparser/package-info.java (hunk 4 above)
Comment (denis-yuen) [reply wdl-parsing/WDLParsingFunction/src/main/java/io/dockstore/wdlparser/package-info.java:2]: Looks like we've been lazy, but there is https://github.com/dockstore/dockstore-documentation/commit/844ffbb51034d46b423a51bd99e6ed7a179a256e

File: wdl-parsing/WDLParsingFunction/src/main/java/io/dockstore/wdlparser/App.java [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.eclipse.jgit.api.Git;
+import org.eclipse.jgit.api.errors.GitAPIException;
+import org.openapitools.client.model.LanguageParsingRequest;
```
Comment (coverbeck) [inline wdl-parsing/WDLParsingFunction/src/main/java/io/dockstore/wdlparser/App.java:38]: It's internal usage only, but it would be nice if a less generic package name were generated.

File: wdl-parsing/WDLParsingFunction/src/main/java/io/dockstore/wdlparser/App.java [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+  ObjectMapper mapper = new ObjectMapper();
+  private static final Logger LOGGER = LoggerFactory.getLogger(App.class);
+
+  @Override
+  public APIGatewayProxyResponseEvent handleRequest(
+      final APIGatewayProxyRequestEvent input, final Context context) {
+
+    Map<String, String> headers = new HashMap<>();
+    headers.put("Content-Type", "application/json");
+    headers.put("X-Custom-Header", "application/json");
```
Comment (coverbeck) [inline wdl-parsing/WDLParsingFunction/src/main/java/io/dockstore/wdlparser/App.java:59]: What is this for?

File: wdl-parsing/WDLParsingFunction/src/main/java/io/dockstore/wdlparser/App.java [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+    }
+  }
+
+  private String parseWdlFile(
+      String uri,
+      String branch,
+      String descriptorRelativePathInGit,
+      LanguageParsingRequest languageParsingRequest)
+      throws IOException, GitAPIException {
+    Path tempDirWithPrefix = Files.createTempDirectory("clonedRepository");
```
Comment (coverbeck) [inline wdl-parsing/WDLParsingFunction/src/main/java/io/dockstore/wdlparser/App.java:106]: Does this need to be deleted?

https://stackoverflow.com/questions/44108712/aws-lambda-release-tmp-storage-after-each-execution, which is a little old, says the disk can get reused.

File: wdl-parsing/WDLParsingFunction/src/main/java/io/dockstore/wdlparser/App.java [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+
+      // The first two lines aren't actual paths.
+      VersionTypeValidation versionTypeValidation = new VersionTypeValidation();
+      if (strings.get(0).equals("Success!")
+          && strings.get(1).equals("List of Workflow dependencies is:")) {
+        versionTypeValidation.setValid(true);
+        response.setVersionTypeValidation(versionTypeValidation);
+        handleSuccessResponse(response, strings);
+      } else {
+        versionTypeValidation.setValid(false);
```
Comment (coverbeck) [inline wdl-parsing/WDLParsingFunction/src/main/java/io/dockstore/wdlparser/App.java:166]: Shouldn't we be capturing the error via `VersionTypeValidation.setMessage`?

File: wdl-parsing/WDLParsingFunction/src/main/java/io/dockstore/wdlparser/App.java [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+        response.setVersionTypeValidation(versionTypeValidation);
+      }
+      return response;
+    } catch (StackOverflowError e) {
+      VersionTypeValidation versionTypeValidation = new VersionTypeValidation();
+      versionTypeValidation.setValid(false);
```
Comment (coverbeck) [inline wdl-parsing/WDLParsingFunction/src/main/java/io/dockstore/wdlparser/App.java:172]: Shouldn't we be capturing the error via `VersionTypeValidation.setMessage`?

File: wdl-parsing/WDLParsingFunction/src/main/java/io/dockstore/wdlparser/App.java (hunk 7 above)
Comment (garyluu) [reply wdl-parsing/WDLParsingFunction/src/main/java/io/dockstore/wdlparser/App.java:59]: default generated

File: wdl-parsing/WDLParsingFunction/src/main/java/io/dockstore/wdlparser/App.java (hunk 9 above)
Comment (garyluu) [reply wdl-parsing/WDLParsingFunction/src/main/java/io/dockstore/wdlparser/App.java:166]: wasn't added in this PR because it's non-trival

File: wdl-parsing/WDLParsingFunction/src/main/java/io/dockstore/wdlparser/App.java (hunk 10 above)
Comment (garyluu) [reply wdl-parsing/WDLParsingFunction/src/main/java/io/dockstore/wdlparser/App.java:172]: wasn't added in this PR because it's non-trival

File: wdl-parsing/WDLParsingFunction/src/main/java/io/dockstore/wdlparser/App.java [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+      List<String> strings = new ArrayList<>();
+      Collections.addAll(strings, split);
+
+      // The first two lines aren't actual paths.
+      VersionTypeValidation versionTypeValidation = new VersionTypeValidation();
+      if (strings.get(0).equals("Success!")
+          && strings.get(1).equals("List of Workflow dependencies is:")) {
+        versionTypeValidation.setValid(true);
+        response.setVersionTypeValidation(versionTypeValidation);
+        handleSuccessResponse(response, strings);
```
Comment (coverbeck) [inline wdl-parsing/WDLParsingFunction/src/main/java/io/dockstore/wdlparser/App.java:84]: `strings` can include fully qualified http(s) imports. Is it your intent that they be part of the response? I can see it either way.

```
java -jar ~/Downloads/womtool-54.jar  validate -l Dockstore.wdl
Success!
List of Workflow dependencies is:
https://raw.githubusercontent.com/coverbeck/bravo-pipeline/noroot/vcfPercentilesPreparation.wdl
https://raw.githubusercontent.com/coverbeck/bravo-pipeline/noroot/cramPreparation.wdl
https://raw.githubusercontent.com/coverbeck/bravo-pipeline/noroot/coveragePreparation.wdl
```

Comment (garyluu) [reply wdl-parsing/WDLParsingFunction/src/main/java/io/dockstore/wdlparser/App.java:84]: yes

**Review by denis-yuen — APPROVED** [review]

File: wdl-parsing/WDLParsingFunction/src/main/java/io/dockstore/wdlparser/App.java [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+        .setBranch(branch)
+        .setURI(uri)
+        .setDirectory(tempDirWithPrefix.toFile())
+        .call();
+    Path descriptorAbsolutePath = tempDirWithPrefix.resolve(descriptorRelativePathInGit);
+    String descriptorAbsolutePathString = descriptorAbsolutePath.toString();
+    LanguageParsingResponse response = getResponse(descriptorAbsolutePathString);
+    response.setLanguageParsingRequest(languageParsingRequest);
+    // Deleting a directory without Common IO
+    Files.walk(tempDirWithPrefix)
```
Comment (coverbeck) [inline wdl-parsing/WDLParsingFunction/src/main/java/io/dockstore/wdlparser/App.java:185]: Could go into a finally, although if the clone fails with an exception, there may not be any files in there.

**Review by coverbeck — APPROVED** [review]
I guess you're not there yet, but deployment is fuzzy to me. Will we deploy using serverless (never done it, seems like you can in conjunction with template.yaml), or do you need to upload a JAR as part of the build which another CF template will deploy?

**Comment by garyluu** [comment]
> I guess you're not there yet, but deployment is fuzzy to me. Will we deploy using serverless (never done it, seems like you can in conjunction with template.yaml), or do you need to upload a JAR as part of the build which another CF template will deploy?

Deployment is pretty much the same as any lambda. there's a template.yaml. It's actually running already running on dockstore/dockstore tests

**Review by NatalieEO — APPROVED** [review]

**Comment by wshands** [comment]
I assume we will also have a Lambda that parses CWL using cwltool?
Can we somehow use this model to parse WDL to create the DAG (not sure if womtool can help with the DAG)

**Review by wshands — APPROVED** [review]
